overlord/snapstate/target.go

```diff
@@ -0,0 +1,588 @@
+			return nil, nil, errors.New("internal error: target cannot specify both a path and a download info")
+		}
+
+		info := inst.info
+
+		if opts.Flags.RequireTypeBase && info.Type() != snap.TypeBase && info.Type() != snap.TypeOS {
+			return nil, nil, fmt.Errorf("unexpected snap type %q, instead of 'base'", info.Type())
+		}
+
+		opts.PrereqTracker.Add(info)
+
+		flags, err := earlyChecks(st, &inst.snapst, info, opts.Flags)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		providerContentAttrs := defaultProviderContentAttrs(st, info, opts.PrereqTracker)
+		snapsup := &SnapSetup{
+			Channel:      inst.setup.Channel,
+			DownloadInfo: inst.setup.DownloadInfo,
+			SnapPath:     inst.setup.SnapPath,
+			CohortKey:    inst.setup.CohortKey,
+
+			Base:               info.Base,
+			Prereq:             getKeys(providerContentAttrs),
+			PrereqContentAttrs: providerContentAttrs,
+			UserID:             opts.UserID,
+			Flags:              flags.ForSnapSetup(),
+			SideInfo:           &info.SideInfo,
+			Type:               info.Type(),
+			Version:            info.Version,
+			PlugsOnly:          len(info.Slots) == 0,
+			InstanceKey:        info.InstanceKey,
+			ExpectedProvenance: info.SnapProvenance,
+		}
+
+		var instFlags int
+		if opts.Flags.SkipConfigure {
+			instFlags |= skipConfigure
+		}
+
+		ts, err := doInstall(st, &inst.snapst, snapsup, instFlags, opts.FromChange, inUseFor(opts.DeviceCtx))
+		if err != nil {
+			return nil, nil, err
+		}
+
+		ts.JoinLane(generateLane(st, opts))
+
+		tasksets = append(tasksets, ts)
+		infos = append(infos, info)
+	}
+
+	return infos, tasksets, nil
+}
+
+func generateLane(st *state.State, opts Options) int {
+	// If transactional, use a single lane for all snaps, so when
+	// one fails the changes for all affected snaps will be
+	// undone. Otherwise, have different lanes per snap so failures
+	// only affect the culprit snap.
+	switch opts.Flags.Transaction {
+	case client.TransactionAllSnaps:
+		return opts.Flags.Lane
+	case client.TransactionPerSnap:
+		return st.NewLane()
+	}
+	return 0
+}
+
+func setDefaultSnapstateOptions(st *state.State, opts *Options) error {
+	var err error
+	if opts.Seed {
+		opts.DeviceCtx, err = DeviceCtxFromState(st, opts.DeviceCtx)
+	} else {
+		opts.DeviceCtx, err = DevicePastSeeding(st, opts.DeviceCtx)
+	}
+	return err
+}
+
+// pathInstallGoal represents a single snap to be installed from a path on disk.
+type pathInstallGoal struct {
+	// path is the path to the snap on disk.
+	path string
+	// instanceName is the name of the snap to install.
+	instanceName string
+	// revOpts contains options that apply to the installation of this snap.
+	revOpts RevisionOptions
+	// sideInfo contains extra information about the snap.
+	sideInfo *snap.SideInfo
+}
+
+// PathInstallGoal creates a new pathInstallGoal from the given instance name,
+// path, and side info.
+func PathInstallGoal(instanceName, path string, si *snap.SideInfo, opts RevisionOptions) InstallGoal {
```

Comment:
the doc comment should explain that if instanceName is ""  si.RealName will be used